cpp/src/sampling/rw_traversals.hpp

```diff
@@ -96,6 +104,240 @@ value_t const* raw_const_ptr(device_const_vector_view<value_t>& dv)
   return dv.begin();
 }
 
+// Uniform RW selector logic:
+//
+template <typename graph_type, typename real_t>
+struct uniform_selector_t {
+  using vertex_t = typename graph_type::vertex_type;
+  using edge_t   = typename graph_type::edge_type;
+  using weight_t = typename graph_type::weight_type;
+
+  struct sampler_t {
+    sampler_t(edge_t const* ro,
+              vertex_t const* ci,
+              weight_t const* w,
+              edge_t const* ptr_d_cache_out_degs)
+      : row_offsets_(ro), col_indices_(ci), values_(w), ptr_d_cache_out_degs_(ptr_d_cache_out_degs)
+    {
+    }
+
+    __device__ thrust::optional<thrust::tuple<vertex_t, weight_t>> operator()(vertex_t src_v,
+                                                                              real_t rnd_val) const
+    {
+      auto crt_out_deg = ptr_d_cache_out_degs_[src_v];
+      if (crt_out_deg == 0) return thrust::nullopt;  // src_v is a sink
+
+      real_t max_ub     = static_cast<real_t>(crt_out_deg - 1);
+      auto interp_vindx = rnd_val * max_ub + real_t{.5};
+      vertex_t v_indx   = static_cast<vertex_t>(interp_vindx);
+
+      auto col_indx  = v_indx >= crt_out_deg ? crt_out_deg - 1 : v_indx;
```

Comment:
Assuming that ```rnd_val``` is a random number in [0, 1.0) and ```crt_out_deg``` is the vertex degree, this code wants to uniformly sample an integer in [0, vertex_degree - 1], right?

Say, the vertex degree is 3. Then, max_ub is 2.0 and interp_vindx will be in [0.5, 2.5). Considering that static_cast truncates, [0.5, 1.0) will be 0, [1.0, 2.0) will be 1, and [2.0, 2.5) will be 2. So, the index 1 has two times higher chance to be selected than 0 or 2, right? Or am I missing something?

Reply:
Same as above:
"I either have to: (a) use floor(), or (b) change the "sanity check" into:

return (v_indx >= crt_out_deg ? 0 : v_indx);

I'll opt for (a), unless a different course of action is suggested."

Reply:
Fixed. (removed ceiling 0.5 logic).